client/rpc-spec-v2/src/chain_head/chain_head_storage.rs

```diff
@@ -110,36 +123,55 @@ where
 
 		result
 			.map(|opt| {
-				opt.map(|storage_data| {
-					QueryResult::Ok(StorageResult::<String> {
-						key: hex_string(&key.0),
-						result: StorageResultType::Hash(hex_string(&storage_data.as_ref())),
-					})
-				})
+				QueryResult::Ok(opt.map(|storage_data| StorageResult::<String> {
+					key: hex_string(&key.0),
+					result: StorageResultType::Hash(hex_string(&storage_data.as_ref())),
+				}))
 			})
 			.unwrap_or_else(|err| {
-				Some(QueryResult::Err(ChainHeadStorageEvent::<String>::Error(ErrorEvent {
+				QueryResult::Err(ChainHeadStorageEvent::<String>::Error(ErrorEvent {
 					error: err.to_string(),
-				})))
+				}))
 			})
 	}
 
-	/// Make the storage query.
-	fn query_storage(
+	/// Handle iterating over (key, value) or (key, hash) pairs.
+	fn query_storage_iter(
 		&self,
 		hash: Block::Hash,
-		query: &StorageQuery<StorageKey>,
+		key: &StorageKey,
 		child_key: Option<&ChildInfo>,
-	) -> Option<QueryResult> {
-		if !is_key_queryable(&query.key.0) {
-			return None
+		ty: IterQueryType,
+	) -> QueryIterResult {
+		let mut keys_iter = if let Some(child_key) = child_key {
+			self.client.child_storage_keys(hash, child_key.to_owned(), Some(key), None)
+		} else {
+			self.client.storage_keys(hash, Some(key), None)
 		}
+		.map_err(|err| {
+			ChainHeadStorageEvent::<String>::Error(ErrorEvent { error: err.to_string() })
+		})?;
+
+		let mut ret = Vec::with_capacity(MAX_ITER_ITEMS);
+		let mut maximum_iters = MAX_ITER_ITEMS;
+		while let Some(key) = keys_iter.next() {
+			if maximum_iters > 0 {
+				maximum_iters -= 1;
+			}
```

Comment:
We would give out more than 10 items here, `maximum_iters` is counting down but does not abort. How about something like this?
```suggestion
		keys_iter = keys_iter.take(MAX_ITER_ITEMS);
		while let Some(key) = keys_iter.next() {
```

Reply:
Yep, that's more elegant! Thanks!